Requirements.md

```diff
@@ -0,0 +1,101 @@
+   
+   Our reasoning here is that performance is often the result of progressive tuning of an implementation. It can typically be improved over time. On the other hand, architectural deficiencies are very difficult to correct.
+      
+1. Performance is more important than convenience.
+
+   Since WASM-GC is not intended as a user-facing technology, convenience here refers to the ease with which toolchain authors can target WASM-GC. 
+   
+   Convenience for toolchain authors is still important however, as that affects adoption of WASM-GC. 
+
+   
+## Critical Success Factors
+
+A critical success factor is an aspect or property of possible solutions that may or may not be directly focused on the primary objective, but none-the-less is estimated to be crucially important to the success of the effort.
+
+1. Permit ‘cycle detection’ between host structures and language structures; so that such cycles can be collected if there are no other references to them.
+
+   When a WASM-GC module accesses host capabilities, or when a host application access structures from a WASM-GC library, cross references between them are likely to be established. For example, a DOM node may have an event listener that is actually implemented in WASM. The event listener, in turn, may have a reference to the DOM node it is listening to. This represents a cycle between the host and WASM-GC. If there are no other references to this cycles (if, for example, the DOM node is removed) then the entire cycle must be able to be collected.
+
+1. Performance implications of WASM-GC code should be straightforward.
+
+   A general principle behind WASM itself is that performance should not require special implementation techniques on the part of the engine. I.e., it should not be required that the engine discover particular opportunities for improving performance. In particular, it is the responsibility of languages implementers to generate WASM-GC code that embodies performant strategies.
+   
+   For example, if a type cast is potentially eliminatable, then the engine should not be required to discover that fact; rather the toolchain should be able to eliminate the type cast.
+   
+1. Support code migration.
+
+   This refers to a potential code migration strategy; where an application starts as a JavaScript code base (say), and is then progressively migrated to another language (hosted on WASM-GC). During this migration process, individual functions and modules may be ported from JavaScript whilst the remaining codebase remains in JavaScript. 
+   Interoperability between WASM-GC code and JavaScript is critical for the success of this strategy; in particular, structures defined in one language may need to be accessed by another.
+   Code Migration itself is a CSF because it underlies a realistic assumption about porting libraries and applications to WASM-GC.
```

Comment:
To make a concrete suggestion addressing all of my concerns:

```suggestion
1. Excellent, fast interoperability with JavaScript and Web APIs

   With JavaScript and Web APIs (the existing Web Platform) as the de facto primary interoperability target of Wasm GC in browsers, designing the interaction between Wasm GC and the existing Web Platform to be frictionless is a CSF as it underlies realistic assumptions about

   * Progressively migrating (parts of) an existing JavaScript codebase to one or multiple (domain-specific) Wasm GC-enabled languages
   * Writing new libraries utilizing Wasm GC-enabled languages to integrate into an existing JavaScript codebase
   * Writing new applications in Wasm GC-enabled languages interacting with Web APIs exclusively
   
   Excellent interoperability between Wasm GC and the Web Platform is critical for the success of these and other strategies and use cases; in particular, strings, arrays, sets, maps and structured objects defined in one language or constructed by a browser may need to be accessed by another language or be consumed by a browser in the absence of sizable and/or inefficient glue code.
```

Reply:
These items are complementary. For example, WebAssembly/design#1345 gave a design that provided better JS interop but at the cost of code migration. That is, if library A and library B spoke to each other, WebAssembly/design#1345 would work fine if just one of the two migrated to WebAssembly, but it broke down if both of them migrated (as in the behavior of the programs composed together changed significantly).

So I would prefer your suggestion as taking the current second bullet of Code Migration and elaborating it into its own item (which seems worthwhile to do).

On a separate note, I'm not really sure what the final sentence phrase (after the semicolon) is saying. It seems like the sort of thing that is more detailed than necessary for this sort of document.

Reply:
Not particularly attached to the exact wording or structure, as long as it still covers excellent and fast interop. I agree that some of this can be considered as too specific, but thought I add these concerns to raise awareness, since these remained unaddressed elsewhere and are relevant. For instance, there's no meaningful interop with Web APIs if there isn't at least a story for strings, and largish or inefficient glue code will never be fast.

Btw I expect WebAssembly/design#1345 to resurface eventually. Unless we decide to cheap out on interop with the Web of course, then the whole ship is going to sink anyway.

Reply:
Sounds good.

> Btw I expect WebAssembly/design#1345 to resurface eventually.

It's on the agenda for tomorrow in fact 😃